### ISSUE TITLE
How to configure a Query based on System.Type

### ISSUE THREAD

**Issue by promontis** [ChilliCream/hotchocolate#1058]
For example, say I have a `Customer` POCO:
```
public class Customer
{
   public string Name { get; set; }
}
```
I can register this via `.AddType(typeof(Customer))`. 
I can then configure the `QueryType` as following:
```
protected override void Configure(IObjectTypeDescriptor<Query> descriptor)
{
	ConfigureType<Customer>(descriptor);			
}

private void ConfigureType<T>(IObjectTypeDescriptor<Query> descriptor)
{
    descriptor.Field(typeof(T).Name).Resolver(() => _resolver.Get<T>());
}
```

This works as long as `T` is one of the (POCO, not `ObjectType`) types I've added in the `SchemaBuilder`.

Since the resolver is typed it can determine the result type of the query to be of `Customer`.
I can now execute the following graphQL query:
```
{
   Customer {
      name
   }
}
```

Now, I would like to resolve my types via reflection.
Given a `List<System.Type>` I can add this to the `SchemaBuilder`.

How to configure the result type of the `descriptor`'s `Resolver` for the `Query` using a `System.Type` instead of it being typed?
I've noticed this overload:
```
IObjectFieldDescriptor Resolver(FieldResolverDelegate fieldResolver, Type resultType)
```

However, it doesn't work for this case specific. 
https://github.com/ChilliCream/hotchocolate/blob/090a84bc9271e2a1bee67f9f22755dd84e4c18c7/src/Core/Types/Types/Descriptors/ObjectFieldDescriptor.cs#L185 throws `InvalidOperationException`, because `Customer` is not a generic type.

Is it currently possible?

**Comment by promontis** [comment]
I seems this overload `IObjectFieldDescriptor Resolver(FieldResolverDelegate fieldResolver, Type resultType)` is working correctly if the `resultType` is `typeof(IEnumerable<Customer>)`.

I've added a PR (#1059) so that we first check to see if the `resultType` is generic to prevent the `InvalidOperationException`

**Comment by michaelstaib** [comment]
Hi @promontis,

thanks for your effort.

The original code seems to always assume that we have a generic type. I have to look at this more closely why this is written like that.

We need some tests that show the error and prove that the fix works here.

I will look at it tomorrow and probably write some tests for this.

**Comment by promontis** [comment]
@michaelstaib that would be awesome! 

I think the original is just a bug, since it does check whether the type is generic or not.

**Comment by promontis** [comment]
@michaelstaib how is it going? Can I help you with anything?

I got a related question for you if possible...

When I try to specify a `.Type()` for say `IArgumentDescriptor` or `IInputFieldDescriptor`, I would like to type it using a `string`. It might sound weird, but I'm 'dynamically' creating the schema for some parts.

I've tried adding something like the following:
```
 public class CustomInputType : InputObjectType
    {
        private readonly Type _forType;

        public CustomInputType(Type forType)
        {
            _forType = forType;
        }

        protected override void Configure(IInputObjectTypeDescriptor descriptor)
        {
            var name = new NameString(_forType.Name);
            descriptor.Name(name);
        }
    }
```

and

```
fieldDescriptor.Argument(new NameString("foo"), a => a.Type(new CustomInputType(type)));
```

And it works, but only when there is just one call to `.Type()` for a specific type name. It seems the `.Type()`is actually adding a type to the Schema and when calling `.Type()` multiple times with the same type name,  say 'Foo', we get the following exception:
```
HotChocolate.SchemaException: The name `Foo` was already registered by another type. - Type: Foo
```

Am I correct that it works this way? If so, why is `.Type<StringType>()` not adding a type (because that seems to work when calling multiple times)? What's the logic behind this?

**Comment by michaelstaib** [comment]
Sorry, I am out with a migraine for the last two days... still not up to 100% so I will look at this on Monday.

**Comment by promontis** [comment]
@michaelstaib no pressure... take your time 🤗

**Comment by michaelstaib** [comment]
@promontis  is your issue solved with the pr?

**Comment by promontis** [comment]
Partly I think, still wondering what the logic is for providing an Argument Type. 
This question... https://github.com/ChilliCream/hotchocolate/issues/1058#issuecomment-528606798

**Comment by michaelstaib** [comment]
did you see that you can pass in a string as name string. so you do not have to create a new name string object. 

So the argument... let`s say you have an argument that you want to pass in your field, then you can declare the argument with that.

Also if the argument was inferred then you can specify more details or override the inferred argument.

you can access the argument through the context ... `ctx.Argument<string>("foo")`.

or if you have a method as resolver:

```csharp
public object Resolver(string foo) 
{
}
```

**Comment by michaelstaib** [comment]
ah... ok so if you do that `a.Type(new CustomInputType(type))` then you register a specific type here. this type can only exist once.... so if you create another instance with the same name then you get an error.

better is to use the generic type. We also have an open issue to have something to have a named reference to a type.

At the moment this can be done to the syntax api which is not so nice but works...

`desc.Type(new ListTypeNode(new NamedTypeNode("Foo"))`

you could then register the type at the schema level.

But there are cases where it is good to register actual type reference in place. We use this with filters or also with paging.

**Comment by michaelstaib** [comment]
I have a view more refinements for the schema builder planned for 10.1.0 that will improve this.

**Comment by michaelstaib** [comment]
The PR is merged.

**Comment by michaelstaib** [comment]
Since the PR was on the master it is now in version 11. I will create a preview in a view minutes. I think I will cherry pick this change tomorrow into the version_10_0_0_master branch so that we get it into the patch build.

**Comment by promontis** [comment]
Yeah, I think the named reference to a type would solve the issue I have. 

I have several hundreds of types in my domain. I could specify them all by hand, but I rather not. 
I now just use reflection to discover the types and that works great. I do need to specify the queries, but they are all the same structure for each type. It works great with the newly merged PR for queries (resultType is the type of the query method), but there currently is no way to do this for arguments. 

Like you said, you can only create new types and do it only once. I would like to create them upfront, and reference them. 

Looking forward to the new features!

We can close this :)

**Comment by michaelstaib** [comment]
> Like you said, you can only create new types and do it only once. I would like to create them upfront, and reference them

This is what you can do now.
`desc.Type(new ListTypeNode(new NamedTypeNode("Foo"))`

We just want to have a nicer syntax for the new version.

Also did you know that you also can just register types and we will infer all the schema stuff.

```csharp
SchemaBuilder.New()
    .AddType(typeof(AnyType))
    .Create();
```

Sometimes people do not realize.

Also the rules on how we infer types can be overridden or modified.
You can even register your own set of attributes etc...

**Comment by michaelstaib** [comment]
You can also extend those types... with type extensions etc...

**Comment by michaelstaib** [comment]
With type extensions you can also split your query types ...

**Comment by michaelstaib** [comment]
https://www.nuget.org/packages/HotChocolate/11.0.0-preview.13

**Comment by michaelstaib** [comment]
TODOs:
- [ ] merge change to 10.1.0

**Comment by promontis** [comment]
@michaelstaib the `11.0.0-preview.13` works great!

> This is what you can do now. `desc.Type(new ListTypeNode(new NamedTypeNode("Foo"))`

That doesn't seem to work for `IArgumentDescriptor`. That class needs a `TInputType where TInputType : class, IInputType`

`ListTypeNode`, nor `NamedTypeNode` is of `IInputType`.

Would it be ok for you if we'd add on `IObjectFieldDescriptor` either
1) `IObjectFieldDescriptor Argument(NameString argumentName, IArgumentDescriptor argumentDescriptor);`. So not an `Action`, but creating it ourselves via the constructor or static methods available
2) `IArgumentDescriptor Type(Type type);` So, without the type constraint `TInputType`, but we should still check whether the type `IsSchemaType` and `!typeof(IInputType).IsAssignableFrom(type)`

Both options would allow me to specify a type via reflection AFAIK and that is what I need.
I prefer the second option.

**Comment by michaelstaib** [comment]
> That doesn't seem to work for IArgumentDescriptor. That class needs a TInputType where TInputType : class, IInputType

https://github.com/ChilliCream/hotchocolate/blob/b98611ad23665b755015c6734ed232dbcce385de/src/Core/Types/Types/Descriptors/Contracts/IArgumentDescriptor.cs#L29

**Comment by michaelstaib** [comment]
OK, I have looked at the APIs more closely and I think I understand what you want now much better.

Two things, I want the API to be consistent so I do not want to just and something extra in one place. So, we should introduce `descriptor.Type(Type type)` everywhere we have these three type overloads including the object field descriptor, interface descriptor, input field descriptor and so on.

I will open a ne issue for that and close this one.

**Comment by michaelstaib** [comment]
The new issue is here #1079 I will write down what we have to do to fulfill this.
